### PR TITLE
Add basic RBAC support

### LINKED ISSUE #299: Version to use
Which version of kubernetes the master branch compliant to? I am running kubernetes cluster via kubeadm and the steps doesn't work with either stable or latest versions. Now I am back at using Kubernetes 1.5.6 version but virt-controller and virt-handler throws these errors.

kubectl logs -f virt-handler-6793p
E0707 13:42:27.509285       1 reflector.go:201] kubevirt.io/kubevirt/pkg/kubecli/controller.go:167: Failed to list *v1.VM: User "system:serviceaccount:default:default" cannot list vms.kubevirt.io in the namespace "default". (get vms.kubevirt.io)

kubectl logs -f virt-controller-3230119827-jjz1z
E0707 11:58:42.154691       1 reflector.go:201] kubevirt.io/kubevirt/pkg/informers/virtinformers.go:90: Failed to list *v1.VM: User "system:serviceaccount:default:default" cannot list vms.kubevirt.io in the namespace "default". (get vms.kubevirt.io)

virsh list doesn't list any VMs. Althought kubectl get vms show test testvm, but there isn't a pod created.

### PR DESCRIPTION
**Pull request by fabiand**
Make sure that KubeVirt can run on Kubernetes instances where RBAC is
enabled. Further give the haproxy admin rights right now, to allow it to
forward all calls to the apiserver.

This does not at all provide a more secure way to run KubeVirt. It's
only purpose at this stage is to allow friction-less deployments on
RBAC enabled Kubernetes clusters.

The depolyment can be properly secured, as soon as we have moved from
our haproxy approach to the API Server aggregation.

Fixes #299, #62 and does some initial work for #301.
Obsoletes #322 .

Signed-off-by: Roman Mohr <rmohr@redhat.com>
Signed-off-by: Fabian Deutsch <fdeutsch@redhat.com>

### CODE REVIEW

**Comment by fabiand** [comment]
I'll provide test results once it's ready.

File: manifests/virt-handler.yaml.in [hunk 1]
```diff
@@ -9,6 +9,7 @@ spec:
       labels:
         daemon: virt-handler
     spec:
+      serviceAccountName: kubevirt-infra
```
Comment (davidvossel) [inline manifests/virt-handler.yaml.in]: virt-handler should have a reduced role that limits it to vm objects with verbs the prevent it from being able to delete anything.

File: manifests/rbac.authorization.k8s.io.yaml.in [hunk 2]
```diff
@@ -0,0 +1,84 @@
+apiVersion: rbac.authorization.k8s.io/v1beta1
```
Comment (davidvossel) [inline manifests/rbac.authorization.k8s.io.yaml.in]: Is there any sort of ordering during deployment time that requires this manifest to be posted before any of the others that depend on the kubevirt-infra account?

**Comment by fabiand** [comment]
The current patchset passes:

```
Ran 31 of 31 Specs in 223.621 seconds
SUCCESS! -- 31 Passed | 0 Failed | 0 Pending | 0 Skipped PASS
```

**Comment by fabiand** [comment]
Added a patch to introduce a read-only/view role, but with this change the tests fail.

**Comment by davidvossel** [comment]
> Added a patch to introduce a read-only/view role, but with this change the tests fail. Will get to it later.

I bet I know why. virt-handler needs to be able to write to the VM object's status field.

**Comment by fabiand** [comment]
That can really well be.

/me was about to look where the update was taking place.

Can we merge the patchset?

File: manifests/rbac.authorization.k8s.io.yaml.in (hunk 2 above)
Comment (rmohr) [reply manifests/rbac.authorization.k8s.io.yaml.in]: Worth investigating.

**Comment by rmohr** [comment]
virt-handler also needs permissions to read PVs, PVCs and Secrets. Would be great if you find the time to investigate that. Otherwise starting with one kubevirt-infra role would work for me. Just make sure to update #301  accordingly.

**Comment by davidvossel** [comment]
virt-handler needs to be able to retrieve secrets as well.  It doesn't need list or watch, just 'get'.

k8s secrets are retrieved for iscsi devices

**Comment by fabiand** [comment]
I'm all in for making the permissions more fine granular - I just see it as my task right now to get the basics in place so that we are deployable on RBACed K8s instances (like OpenShift) to unblock others.

Once tihs works we can make this more detailed.

Especially because we can then run automation on those other K8s instances as well ...

**Comment by davidvossel** [comment]
> I'm all in for making the permissions more fine granular - I just see it as my task right now to get the basics in place so that we are deployable on RBACed K8s instances (like OpenShift) to unblock others.

that makes sense. If the broad RBAC role worked, lets just go back to that single role and get this merged for now then.

**Comment by fabiand** [comment]
Rebased, should be good to merge.

**Comment by cynepco3hahue** [comment]
@fabiand Does not help to run privileged pods, for example on libvirt I have:
```
Sep 12 16:06:52 alukiano-openshift-master origin-master-controllers: I0912 16:06:52.531545    6383 event.go:217] Event(v1.ObjectReference{Kind:"DaemonSet", Namespace:"kubevirt", Name:"libvirt", UID:"dce9ff70-97ba-11e7-bdea-001a4a1610c1", APIVersion:"extensions", ResourceVersion:"610702", FieldPath:""}): type: 'Warning' reason: 'FailedCreate' Error creating: pods "libvirt-" is forbidden: unable to validate against any security context constraint: [provider restricted: .spec.securityContext.hostNetwork: Invalid value: true: Host network is not allowed to be used provider restricted: .spec.securityContext.hostPID: Invalid value: true: Host PID is not allowed to be used provider restricted: .spec.securityContext.hostIPC: Invalid value: true: Host IPC is not allowed to be used securityContext.runAsUser: Invalid value: 0: UID on container libvirtd does not match required range.  Found 0, required min: 1000070000 max: 1000079999 provider restricted: .spec.containers[0].securityContext.privileged: Invalid value: true: Privileged containers are not allowed provider restricted: .spec.containers[0].securityContext.volumes[0]: Invalid value: "hostPath": hostPath volumes are not allowed to be used provider restricted: .spec.containers[0].securityContext.volumes[1]: Invalid value: "hostPath": hostPath volumes are not allowed to be used provider restricted: .spec.containers[0].securityContext.volumes[2]: Invalid value: "hostPath": hostPath volumes are not allowed to be used provider restricted: .spec.containers[0].securityContext.volumes[3]: Invalid value: "hostPath": hostPath volumes are not allowed to be used provider restricted: .spec.containers[0].securityContext.volumes[4]: Invalid value: "hostPath": hostPath volumes are not allowed to be used provider restricted: .spec.containers[0].securityContext.hostNetwork: Invalid value: true: Host network is not allowed to be used provider restricted: .spec.containers[0].securityContext.containers[0].hostPort: Invalid value: 16509: Host ports are not allowed to be used provider restricted: .spec.containers[0].securityContext.hostPID: Invalid value: t
```

**Comment by fabiand** [comment]
@cynepco3hahue yes.

This set of patches was just a start. I believe that openshift needs additional changes.
One idea was to add the service account to the cluster role to give it enough privileges.

But I am not sure if that is enough.

**Comment by rmohr** [comment]
@admiyo solved issues with the OpenShift deployment by moving our kubevirt deployment to the `kube-system` and adjusting the OpenShift `PodSecurityPolicy` (https://kubernetes.io/docs/concepts/policy/pod-security-policy/). @cynepco3hahue the ones you mention look like it will be solved by adjusting a PodSecurityPolicy.

His post where he explains everything he had to modify: http://adam.younglogic.com/2017/08/deploying-kubevirt-on-origin-master/

Also note, that he used OpenShift from latest master.

**Comment by cynepco3hahue** [comment]
@rmohr Yes I know, I just thought this patch will fix the problem with privileged user :smile:

**Pull request merged**